File: flink-demo/flinkScalaCode/src/main/scala/com/yw/flink/example/scalacases/case00/Case01_BatchWordCount.scala
package com.yw.flink.example.scalacases.case00

import org.apache.flink.api.scala.ExecutionEnvironment

/**
  * Flink 批处理 Scala WordCount
  *
  * @author yangwei
  */
object Case01_BatchWordCount {
  def main(args: Array[String]): Unit = {
    // 1. 准备环境
    val env: ExecutionEnvironment = ExecutionEnvironment.getExecutionEnvironment
    // 2. 导入隐式转换
    import org.apache.flink.api.scala._
    // 3. 读取数据文件
    val linesDs: DataSet[String] = env.readTextFile(".data/words.txt")
    // 4. 切分单词
    val wordsDs: DataSet[String] = linesDs.flatMap(line => line.split(" "))
    // 5. 对数据进行计数、分组、聚合统计
    wordsDs.map(word => {
      (word, 1)
    }).groupBy(0)
      .sum(1)
      .print()
  }
}

File: flink-demo/flinkJavaCode/src/main/java/com/yw/flink/example/javacases/case00/Case01_BatchWordCount.java
package com.yw.flink.example.javacases.case00;

import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.api.java.ExecutionEnvironment;
import org.apache.flink.api.java.operators.DataSource;
import org.apache.flink.api.java.operators.FlatMapOperator;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.util.Collector;

/**
 * Flink 批处理 WordCount
 *
 * @author yangwei
 */
public class Case01_BatchWordCount {
    public static void main(String[] args) throws Exception {
        // 1. 准备Flink运行环境
        ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();

        // 2. 读取数据文件
        DataSource<String> lineDs = env.readTextFile(".data/words.txt");

        // 3. 对数据进行切分单词，并转成Tuple2 KV 类型
        FlatMapOperator<String, Tuple2<String, Long>> kvWordsDs =
                lineDs.flatMap((String line, Collector<Tuple2<String, Long>> collector) -> {
                    for (String word : line.split(" ")) {
                        collector.collect(Tuple2.of(word, 1L));
                    }
                }).returns(Types.TUPLE(Types.STRING, Types.LONG));

        // 4. 进行数据分组，计数，打印结果
        kvWordsDs.groupBy(0).sum(1).print();
    }
}
